daemon/container_operations_unix.go

```diff
@@ -618,16 +618,39 @@ func (daemon *Daemon) allocateNetwork(container *container.Container) error {
 	}
 
 	updateSettings := false
+	if container.Config.NetworkDisabled || container.HostConfig.NetworkMode.IsContainer() {
+		return nil
+	}
 	if len(container.NetworkSettings.Networks) == 0 {
-		if container.Config.NetworkDisabled || container.HostConfig.NetworkMode.IsContainer() {
-			return nil
-		}
-
 		err := daemon.updateContainerNetworkSettings(container, nil)
 		if err != nil {
 			return err
 		}
 		updateSettings = true
+	} else {
+		networkName := container.HostConfig.NetworkMode.NetworkName()
+		if container.HostConfig.NetworkMode.IsDefault() {
+			networkName = daemon.netController.Config().Daemon.DefaultNetwork
+		}
+		if container.HostConfig.NetworkMode.IsUserDefined() {
+			n, err := daemon.FindNetwork(networkName)
+			if err != nil {
+				return err
+			}
+			networkName = n.Name()
+		}
+		if _, ok := container.NetworkSettings.Networks[networkName]; !ok {
```

Comment:
I think this will break in the following scenario (but not sure how valid is the scenario though) : 

```
docker network create xxx
docker network create yyy
docker network create zzz
docker run --net=xxx --name=foo ...
docker network connect yyy foo
docker network connect zzz foo
docker network disconnect xxx foo
docker restart foo
```
This will result in the container losing connectivity to networks yyy and zzz.

I have to think through other cases or possible fix. But am not even sure if we have to handle this case.